lib/pacemaker/pcmk_sched_location.c

```diff
@@ -92,6 +57,113 @@ parse_location_role(const char *role_spec, enum rsc_role_e *role)
+    if (name == NULL) {
+        return NULL;
+    }
+
+    /* A score attribute name may use submatches extracted from a
+     * resource ID regular expression. For example, if score-attribute is
+     * "loc-\1", rsc-pattern is "ip-(.*)", and the resource ID is "ip-db", then
+     * the score attribute name is "loc-db".
+     */
+    if ((rule_input->rsc_id != NULL) && (rule_input->rsc_id_nmatches > 0)) {
+        *allocated = pcmk__replace_submatches(name, rule_input->rsc_id,
+                                              rule_input->rsc_id_submatches,
+                                              rule_input->rsc_id_nmatches);
+        if (*allocated != NULL) {
+            name = *allocated;
+        }
+    }
+    return name;
+}
+
+/*!
+ * \internal
+ * \brief Parse a score from a rule without a score attribute
+ *
+ * \param[in]  rule_xml    Rule XML
+ * \param[out] score       Where to store parsed score
+ *
+ * \return Standard Pacemaker return code
+ */
+static int
+score_from_rule(const xmlNode *rule_xml, int *score)
+{
+    int rc = pcmk_rc_ok;
+    const char *score_s = crm_element_value(rule_xml, PCMK_XA_SCORE);
+
+    if (score_s == NULL) { // Not possible with schema validation enabled
+        pcmk__config_err("Ignoring location constraint rule %s because "
+                         "neither " PCMK_XA_SCORE " nor "
+                         PCMK_XA_SCORE_ATTRIBUTE " was specified",
+                         pcmk__xe_id(rule_xml));
+        return pcmk_rc_unpack_error;
+    }
+
+    rc = pcmk_parse_score(score_s, score, 0);
+    if (rc != pcmk_rc_ok) { // Not possible with schema validation enabled
+        pcmk__config_err("Ignoring location constraint rule %s because "
+                         "'%s' is not a valid " PCMK_XA_SCORE ": %s",
+                         pcmk__xe_id(rule_xml), score_s, pcmk_rc_str(rc));
+        return pcmk_rc_unpack_error;
+    }
+
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Get a rule score from a node attribute
+ *
+ * \param[in] rule       Rule ID (for logging only)
+ * \param[in] attr_name  Name of node attribute with score
+ * \param[in] node       Node to get attribute for
+ * \param[in] rsc        Resource being located
+ *
+ * \return Score parsed from value of \p attr_name on \p node (default 0)
+ */
+static int
+score_from_attr(const char *rule, const char *attr_name,
+                const pcmk_node_t *node, const pcmk_resource_t *rsc)
+{
+    int score = 0;
+    int rc = pcmk_rc_ok;
+    const char *target = NULL;
+    const char *score_s = NULL;
+
+    target = g_hash_table_lookup(rsc->priv->meta,
+                                 PCMK_META_CONTAINER_ATTRIBUTE_TARGET);
+    score_s = pcmk__node_attr(node, attr_name, target, pcmk__rsc_node_current);
+
+    rc = pcmk_parse_score(score_s, &score, -PCMK_SCORE_INFINITY);
+    if (rc != pcmk_rc_ok) {
+        crm_warn("Using " PCMK_VALUE_MINUS_INFINITY " as score for rule %s "
```

Comment:
I don't really understand this decision to use `-INFINITY` instead of using 0 or ignoring the constraint, if the source of the score is missing or has an invalid value. But it's not wrong.

Reply:
Here, I just wanted invalid values to be treated the same as missing.

I think it does make more sense to ignore the constraint for that node; I'll add a commit for that. I don't think 0 makes sense because that would enable resources on nodes without the attribute in asymmetric clusters.